File: codes/python/chap06/answer_A39.py
# 入力
# A は (終了時刻, 開始時刻) になっていることに注意 [終了時刻の昇順にソートするため]
N = int(input())
A = []
for i in range(N):
	l, r = map(int, input().split())
	A.append([r, l])

# ソート
A.sort()

# 終了時刻の早いものから貪欲に取っていく（CurrentTime は現在時刻）
CurrentTime = 0
Answer = 0
for i in range(N):
	if CurrentTime <= A[i][1]:
		CurrentTime = A[i][0]
		Answer += 1

# 出力
print(Answer)

File: codes/cpp/chap06/answer_A39.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int N, L[300009], R[300009];
vector<pair<int, int>> tmp; // 映画をソートするための一時変数

int main() {
	// 入力
	cin >> N;
	for (int i = 1; i <= N; i++) {
		cin >> L[i] >> R[i];
		tmp.push_back(make_pair(R[i], L[i]));
	}

	// R の小さい順にソート
	sort(tmp.begin(), tmp.end());
	for (int i = 1; i <= N; i++) {
		R[i] = tmp[i - 1].first;
		L[i] = tmp[i - 1].second;
	}

	// 終了時刻の早いものから貪欲に取っていく（CurrentTime は現在時刻）
	int CurrentTime = 0, Answer = 0;
	for (int i = 1; i <= N; i++) {
		if (CurrentTime <= L[i]) {
			CurrentTime = R[i];
			Answer += 1;
		}
	}
	cout << Answer << endl;
	return 0;
}
